vital/tests/test_const_iterator.cxx

```diff
@@ -0,0 +1,409 @@
+ * modification, are permitted provided that the following conditions are met:
+ *
+ *  * Redistributions of source code must retain the above copyright notice,
+ *    this list of conditions and the following disclaimer.
+ *
+ *  * Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ *  * Neither name of Kitware, Inc. nor the names of any contributors may be used
+ *    to endorse or promote products derived from this software without specific
+ *    prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHORS OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <iostream>
+
+#include <gtest/gtest.h>
+
+#include <vital/iterator.h>
+
+using namespace kwiver;
+
+// ----------------------------------------------------------------------------
+int main( int argc, char** argv )
+{
+  ::testing::InitGoogleTest( &argc, argv );
+  return RUN_ALL_TESTS();
+}
+
+// Test functions -------------------------------------------------------------
+
+// Test default construction - const
+TEST( const_iterator, construct_default )
+{
+  vital::const_iterator<int> i;
+}
+
+// Test construction passing a generator function
+TEST( const_iterator, construct_with_generator )
+{
+  typedef vital::const_iterator<int> iter_t;
+
+  iter_t::next_value_func_t i_gen = []()->iter_t::reference{
+    static int v = 0;
+    return v;
+  };
+
+  iter_t i( i_gen );
+}
+
+// Test copy construction
+TEST( const_iterator, construct_copy )
+{
+  typedef vital::const_iterator<int> iter_t;
+  iter_t i1;
+  iter_t i2( i1 );
+}
+
+// Test value assignment
+TEST( const_iterator, assignment )
+{
+  vital::const_iterator<int> i1;
+  i1 = vital::const_iterator<int>();
+  // Iteration type enforced at compile time. The following should fail:
+  //i1 = vital::const_iterator<double>();
+}
+
+// Test prefix incrementing operator over integer sequence generator.
+TEST( const_iterator, prefix_increment )
+{
+  typedef vital::const_iterator<unsigned int> iter_t;
+
+  unsigned int v = -1;
+  iter_t::next_value_func_t nvf = [&]()->iter_t::reference{
+    ++v;
+    return v;
+  };
+
+  iter_t it( nvf );
+  ASSERT_EQ( *it, 0 );
```

Comment:
There are exactly *zero* non-fatal assertions in this entire PR. That is almost certainly not correct. Please check carefully if *any* of these really need to be fatal assertions. Please also keep in mind the immediately-invoked lambda trick (see e.g. `test_geo_polygon.cxx:149`) that allows skipping a block of following assertions without aborting the entire test case.

Reply:
Switching over to ``EXPECT_*`` macros.

Reply:
Done.